File: sdk/typescript/src/infrastructure/chain-client.ts
export interface AccountInfo {
  account_number: number;
  sequence: number;
}

export interface BroadcastResponse {
  tx_hash: string;
  code: number;
  raw_log: string;
}

export interface BlockResponse {
  header: {
    height: number;
    time: string;
    chain_id: string;
  };
  data: {
    txs: Uint8Array[];
  };
}

export interface TxResponse {
  tx_hash: string;
  height: number;
  tx_result: {
    code: number;
    gas_used: number;
    gas_wanted: number;
    log: string;
    events: Array<{
      type: string;
      attributes: Array<{ key: string; value: string }>;
    }>;
    msg_responses?: Array<Record<string, unknown>>;
  };
}

export interface ChainClient {
  connect(rpcEndpoint: string, grpcEndpoint: string): Promise<void>;
  disconnect(): Promise<void>;
  isConnected(): boolean;

  queryRest<T>(path: string): Promise<T>;
  broadcastTx(txBytes: Uint8Array, mode: string): Promise<BroadcastResponse>;
  getAccountInfo(address: string): Promise<AccountInfo>;

  getLatestBlock(): Promise<BlockResponse>;
  getBlockByHeight(height: number): Promise<BlockResponse>;
  getTx(txHash: string): Promise<TxResponse | null>;
}

export class HttpChainClient implements ChainClient {
  private rpcEndpoint: string = "";
  private grpcEndpoint: string = "";
  private connected: boolean = false;

  async connect(rpcEndpoint: string, grpcEndpoint: string): Promise<void> {
    this.rpcEndpoint = rpcEndpoint;
    this.grpcEndpoint = grpcEndpoint;
    this.connected = true;
  }

  async disconnect(): Promise<void> {
    this.connected = false;
    this.rpcEndpoint = "";
    this.grpcEndpoint = "";
  }

  isConnected(): boolean {
    return this.connected;
  }

  async queryRest<T>(path: string): Promise<T> {
    this.ensureConnected();
    const url = `${this.rpcEndpoint}${path}`;
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${response.statusText}`);
    }
    return response.json() as Promise<T>;
  }

  async broadcastTx(
    txBytes: Uint8Array,
    mode: string,
  ): Promise<BroadcastResponse> {
    this.ensureConnected();
    const txBytesBase64 = uint8ArrayToBase64(txBytes);
    const broadcastMode =
      mode === "async" ? "BROADCAST_MODE_ASYNC" : "BROADCAST_MODE_SYNC";

    const url = `${this.rpcEndpoint}/cosmos/tx/v1beta1/txs`;
    const response = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        tx_bytes: txBytesBase64,
        mode: broadcastMode,
      }),
    });

    if (!response.ok) {
      const body = await response.text();
      throw new Error(`broadcast failed: HTTP ${response.status}: ${body}`);
    }

    const result = (await response.json()) as {
      tx_response: {
        txhash: string;
        code: number;
        raw_log: string;
      };
    };

    return {
      tx_hash: result.tx_response.txhash,
      code: result.tx_response.code,
      raw_log: result.tx_response.raw_log,
    };
  }

  async getAccountInfo(address: string): Promise<AccountInfo> {
    this.ensureConnected();
    const url = `${this.rpcEndpoint}/cosmos/auth/v1beta1/accounts/${address}`;
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(
        `account info query failed: HTTP ${response.status}: ${response.statusText}`,
      );
    }

    const result = (await response.json()) as {
      account: {
        account_number?: string;
        sequence?: string;
        base_account?: {
          account_number: string;
          sequence: string;
        };
      };
    };

    // Handle both direct accounts and nested base_account (e.g. vesting accounts)
    const acct = result.account.base_account ?? result.account;
    return {
      account_number: parseInt(acct.account_number ?? "0", 10),
      sequence: parseInt(acct.sequence ?? "0", 10),
    };
  }

  async getLatestBlock(): Promise<BlockResponse> {
    this.ensureConnected();
    const url = `${this.rpcEndpoint}/cosmos/base/tendermint/v1beta1/blocks/latest`;
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(
        `latest block query failed: HTTP ${response.status}: ${response.statusText}`,
      );
    }

    const result = (await response.json()) as {
      block: {
        header: { height: string; time: string; chain_id: string };
        data: { txs: string[] | null };
      };
    };

    return {
      header: {
        height: parseInt(result.block.header.height, 10),
        time: result.block.header.time,
        chain_id: result.block.header.chain_id,
      },
      data: {
        txs: (result.block.data.txs ?? []).map(base64ToUint8Array),
      },
    };
  }

  async getBlockByHeight(height: number): Promise<BlockResponse> {
    this.ensureConnected();
    const url = `${this.rpcEndpoint}/cosmos/base/tendermint/v1beta1/blocks/${height}`;
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(
        `block query failed: HTTP ${response.status}: ${response.statusText}`,
      );
    }

    const result = (await response.json()) as {
      block: {
        header: { height: string; time: string; chain_id: string };
        data: { txs: string[] | null };
      };
    };

    return {
      header: {
        height: parseInt(result.block.header.height, 10),
        time: result.block.header.time,
        chain_id: result.block.header.chain_id,
      },
      data: {
        txs: (result.block.data.txs ?? []).map(base64ToUint8Array),
      },
    };
  }

  async getTx(txHash: string): Promise<TxResponse | null> {
    this.ensureConnected();
    const url = `${this.rpcEndpoint}/cosmos/tx/v1beta1/txs/${txHash}`;
    const response = await fetch(url);

    if (!response.ok) {
      if (response.status === 404) return null;
      const body = await response.text();
      if (body.includes("not found") || body.includes("not exist")) {
        return null;
      }
      throw new Error(
        `tx query failed: HTTP ${response.status}: ${response.statusText}`,
      );
    }

    const result = (await response.json()) as {
      tx_response: {
        txhash: string;
        height: string;
        code: number;
        gas_used: string;
        gas_wanted: string;
        raw_log: string;
        events: Array<{
          type: string;
          attributes: Array<{ key: string; value: string }>;
        }>;
      };
    };

    return {
      tx_hash: result.tx_response.txhash,
      height: parseInt(result.tx_response.height, 10),
      tx_result: {
        code: result.tx_response.code,
        gas_used: parseInt(result.tx_response.gas_used, 10),
        gas_wanted: parseInt(result.tx_response.gas_wanted, 10),
        log: result.tx_response.raw_log,
        events: result.tx_response.events ?? [],
      },
    };
  }

  getRpcEndpoint(): string {
    return this.rpcEndpoint;
  }

  getGrpcEndpoint(): string {
    return this.grpcEndpoint;
  }

  private ensureConnected(): void {
    if (!this.connected) {
      throw new Error("ChainClient is not connected");
    }
  }
}

function uint8ArrayToBase64(bytes: Uint8Array): string {
  if (typeof Buffer !== "undefined") {
    return Buffer.from(bytes).toString("base64");
  }
  let binary = "";
  for (let i = 0; i < bytes.length; i++) {
    binary += String.fromCharCode(bytes[i]);
  }
  return btoa(binary);
}

function base64ToUint8Array(b64: string): Uint8Array {
  if (typeof Buffer !== "undefined") {
    return new Uint8Array(Buffer.from(b64, "base64"));
  }
  const binary = atob(b64);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes;
}

File: sdk/csharp/src/Infrastructure/ChainClient.cs
using System.Text;
using System.Text.Json;
using Seocheon.Sdk.Errors;
using Seocheon.Sdk.Internal.Tx;

namespace Seocheon.Sdk.Infrastructure;

/// <summary>
/// Chain client interface for REST queries and TX broadcast.
/// </summary>
public interface IChainClient
{
    /// <summary>Establishes connection to the chain.</summary>
    Task ConnectAsync(CancellationToken ct = default);

    /// <summary>Closes the connection.</summary>
    Task DisconnectAsync();

    /// <summary>Returns whether the client is connected.</summary>
    bool IsConnected { get; }

    /// <summary>Queries a REST endpoint.</summary>
    Task<JsonElement> QueryRest(string path, CancellationToken ct = default);

    /// <summary>Broadcasts a signed transaction.</summary>
    Task<BroadcastResponse> BroadcastTx(byte[] txBytes, string mode, CancellationToken ct = default);

    /// <summary>Gets the latest block info.</summary>
    Task<JsonElement> GetLatestBlock(CancellationToken ct = default);

    /// <summary>Queries a transaction by hash.</summary>
    Task<TxResponse?> GetTx(string txHash, CancellationToken ct = default);

    /// <summary>Gets account info (number + sequence).</summary>
    Task<AccountInfo> GetAccountInfo(string address, CancellationToken ct = default);
}

/// <summary>
/// HTTP-based chain client implementation.
/// </summary>
public sealed class HttpChainClient : IChainClient, IDisposable
{
    private readonly string _rpcEndpoint;
    private readonly string _grpcEndpoint;
    private HttpClient? _http;
    private bool _connected;

    /// <summary>
    /// Creates an HttpChainClient for the given endpoints.
    /// </summary>
    public HttpChainClient(string rpcEndpoint, string grpcEndpoint)
    {
        _rpcEndpoint = rpcEndpoint.TrimEnd('/');
        _grpcEndpoint = grpcEndpoint.TrimEnd('/');
    }

    /// <inheritdoc />
    public bool IsConnected => _connected;

    /// <inheritdoc />
    public Task ConnectAsync(CancellationToken ct = default)
    {
        _http = new HttpClient { Timeout = TimeSpan.FromSeconds(30) };
        _connected = true;
        return Task.CompletedTask;
    }

    /// <inheritdoc />
    public Task DisconnectAsync()
    {
        _http?.Dispose();
        _http = null;
        _connected = false;
        return Task.CompletedTask;
    }

    /// <inheritdoc />
    public async Task<JsonElement> QueryRest(string path, CancellationToken ct = default)
    {
        EnsureConnected();
        try
        {
            var url = $"{_grpcEndpoint}{path}";
            var response = await _http!.GetAsync(url, ct);

            if (!response.IsSuccessStatusCode)
                throw SdkErrors.QueryFailed($"HTTP {response.StatusCode}: {path}");

            var body = await response.Content.ReadAsStringAsync(ct);
            return JsonDocument.Parse(body).RootElement.Clone();
        }
        catch (SdkException)
        {
            throw;
        }
        catch (Exception ex)
        {
            throw SdkErrors.QueryFailed($"{path}: {ex.Message}");
        }
    }

    /// <inheritdoc />
    public async Task<BroadcastResponse> BroadcastTx(byte[] txBytes, string mode, CancellationToken ct = default)
    {
        EnsureConnected();
        try
        {
            var payload = JsonSerializer.Serialize(new
            {
                tx_bytes = Convert.ToBase64String(txBytes),
                mode = mode == "async" ? "BROADCAST_MODE_ASYNC" : "BROADCAST_MODE_SYNC"
            });

            var content = new StringContent(payload, Encoding.UTF8, "application/json");
            var response = await _http!.PostAsync(
                $"{_grpcEndpoint}/cosmos/tx/v1beta1/txs",
                content,
                ct
            );

            var body = await response.Content.ReadAsStringAsync(ct);
            using var doc = JsonDocument.Parse(body);
            var txResponse = doc.RootElement.GetProperty("tx_response");

            return new BroadcastResponse
            {
                TxHash = txResponse.GetProperty("txhash").GetString() ?? "",
                Code = txResponse.GetProperty("code").GetUInt32(),
                RawLog = txResponse.TryGetProperty("raw_log", out var rl) ? rl.GetString() ?? "" : ""
            };
        }
        catch (SdkException)
        {
            throw;
        }
        catch (Exception ex)
        {
            throw SdkErrors.BroadcastFailed(ex.Message);
        }
    }

    /// <inheritdoc />
    public async Task<JsonElement> GetLatestBlock(CancellationToken ct = default)
    {
        return await QueryRest("/cosmos/base/tendermint/v1beta1/blocks/latest", ct);
    }

    /// <inheritdoc />
    public async Task<TxResponse?> GetTx(string txHash, CancellationToken ct = default)
    {
        EnsureConnected();
        try
        {
            var response = await _http!.GetAsync(
                $"{_grpcEndpoint}/cosmos/tx/v1beta1/txs/{txHash}",
                ct
            );

            if (!response.IsSuccessStatusCode)
                return null;

            var body = await response.Content.ReadAsStringAsync(ct);
            using var doc = JsonDocument.Parse(body);
            var txResp = doc.RootElement.GetProperty("tx_response");

            var events = new List<TxEventData>();
            if (txResp.TryGetProperty("events", out var eventsEl))
            {
                foreach (var evt in eventsEl.EnumerateArray())
                {
                    var attrs = new Dictionary<string, string>();
                    if (evt.TryGetProperty("attributes", out var attrsEl))
                    {
                        foreach (var attr in attrsEl.EnumerateArray())
                        {
                            var key = attr.GetProperty("key").GetString() ?? "";
                            var val = attr.TryGetProperty("value", out var v) ? v.GetString() ?? "" : "";
                            attrs[key] = val;
                        }
                    }
                    events.Add(new TxEventData(
                        evt.GetProperty("type").GetString() ?? "",
                        attrs
                    ));
                }
            }

            return new TxResponse
            {
                TxHash = txResp.GetProperty("txhash").GetString() ?? "",
                Height = long.Parse(txResp.GetProperty("height").GetString() ?? "0"),
                Code = txResp.GetProperty("code").GetUInt32(),
                RawLog = txResp.TryGetProperty("raw_log", out var rl) ? rl.GetString() ?? "" : "",
                Events = events
            };
        }
        catch (SdkException)
        {
            throw;
        }
        catch
        {
            return null;
        }
    }

    /// <inheritdoc />
    public async Task<AccountInfo> GetAccountInfo(string address, CancellationToken ct = default)
    {
        try
        {
            var result = await QueryRest($"/cosmos/auth/v1beta1/accounts/{address}", ct);
            var account = result.GetProperty("account");

            ulong accountNumber = 0;
            ulong sequence = 0;

            if (account.TryGetProperty("account_number", out var an))
                ulong.TryParse(an.GetString(), out accountNumber);
            if (account.TryGetProperty("sequence", out var seq))
                ulong.TryParse(seq.GetString(), out sequence);

            // Handle nested base_account (vesting accounts)
            if (account.TryGetProperty("base_account", out var baseAcct))
            {
                if (baseAcct.TryGetProperty("account_number", out var ban))
                    ulong.TryParse(ban.GetString(), out accountNumber);
                if (baseAcct.TryGetProperty("sequence", out var bseq))
                    ulong.TryParse(bseq.GetString(), out sequence);
            }

            return new AccountInfo { AccountNumber = accountNumber, Sequence = sequence };
        }
        catch (SdkException)
        {
            throw;
        }
        catch (Exception ex)
        {
            throw SdkErrors.AccountNotFound($"{address}: {ex.Message}");
        }
    }

    /// <summary>Disposes the HTTP client.</summary>
    public void Dispose()
    {
        _http?.Dispose();
    }

    private void EnsureConnected()
    {
        if (!_connected || _http == null)
            throw SdkErrors.NotConnected();
    }
}
